### PR TITLE
p-867 Refactoring extra params

### PR DESCRIPTION
**Pull request by 0xverin**
### Context

This PR refactoring the structure of the contract, using params as parameters to match different token library.

~~We don't need to deploy each token now, just deploy the [ERC20Mapping](https://github.com/litentry/litentry-parachain/pull/2830/files#diff-d55cbd2a8d37a86353163ab51eaf135c93b65a4713046e56952c26337bfb311e) and [BRC20Mapping](https://github.com/litentry/litentry-parachain/pull/2830/files#diff-d9754d115d0f67f4b3e97b3c2eeba4ceb0485bdb90ee80a6fcdf0f657a9eb106) contracts,according to the decoded params,we can find the corresponding mapping.~~

Now just need to deploy the [TokenMapping](https://github.com/litentry/litentry-parachain/pull/2830/files#diff-ffbd58367b287b2fd9014f61f6ccffbfd6e84b451e80279d1d869ce1f5ebe985) contract.

`extra params` should be abi.encode of the token lowercase name, for example:

abi.encode('bnb'):`00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003626e620000000000000000000000000000000000000000000000000000000000`

### Labels

Please apply following PR-related labels when appropriate:
- `C0-breaking`: if your change could break the existing client, e.g. API change, critical logic change 
- `C1-noteworthy`: if your change is non-breaking, but is still worth noticing for the client, e.g. reference code improvement

### Testing Evidences

```
 ./bin/litentry-cli trusted -d link-identity did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e did:litentry:evm:0x5ad727D30bbcB5A7bCdFa213B940cCFA61835121 ethereum 
```

```
./bin/litentry-cli trusted -d request-vc did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e -a "dynamic 0000000000000000000000000000000000000007 00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003626e620000000000000000000000000000000000000000000000000000000000"
```

```
{
    "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/suites/ed25519-2020/v1"
    ],
    "id": "0xd250829993d57dd45d51c6828fde7af1b15c833621672a7cb7a2e63530aa764b",
    "type": [
        "VerifiableCredential"
    ],
    "credentialSubject": {
        "id": "did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e",
        "description": "The amount of a particular token you are holding",
        "type": "Token Holding Amount",
        "assertionText": "Dynamic(0x0000000000000000000000000000000000000007, BoundedVec([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 98, 110, 98, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 1024))",
        "assertions": [
            {
                "and": [
                    {
                        "src": "$token",
                        "op": "==",
                        "dst": "Bnb"
                    },
                    {
                        "src": "$holding_amount",
                        "op": ">=",
                        "dst": "500"
                    },
                    {
                        "src": "$holding_amount",
                        "op": "<",
                        "dst": "800"
                    }
                ]
            }
        ],
        "values": [
            true
        ],
        "endpoint": "http://localhost:9933/"
    },
    "issuer": {
        "id": "did:litentry:substrate:0x7ccfdef3e7cd7bc5bac3b5845095e13c98f7eee5fd481884c7398d877d792b15",
        "name": "Litentry TEE Worker",
        "mrenclave": "9vnN35nXEq76rUS2ywFAgAjjh2HHCF3LjHck9SSuQEwH",
        "runtimeVersion": {
            "parachain": 9181,
            "sidechain": 107
        }
    },
    "issuanceDate": "2024-06-20T05:04:47.349866428+00:00",
    "parachainBlockNumber": 5222,
    "sidechainBlockNumber": 10432,
    "proof": {
        "created": "2024-06-20T05:04:47.351085789+00:00",
        "type": "Ed25519Signature2020",
        "proofPurpose": "assertionMethod",
        "proofValue": "1d851ab95493798f0080157776c24051ebdacdaf1734e1ec386a3614dab2ff3c19744bebb575a98aaa538acaf89087af4434cd7c4ae05fe10065f9119410820b",
        "verificationMethod": "0x7ccfdef3e7cd7bc5bac3b5845095e13c98f7eee5fd481884c7398d877d792b15"
    },
    "credentialSchema": {
        "id": "https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-0.json",
        "type": "JsonSchemaValidator2018"
    }
}
```

image

### CODE REVIEW

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol [hunk 1]
```diff
@@ -21,62 +21,86 @@ pragma solidity ^0.8.8;
 import "../libraries/Identities.sol";
 import "../libraries/Http.sol";
 import "../libraries/Utils.sol";
-import {TokenHoldingAmount} from "./TokenHoldingAmount.sol";
+import { TokenHoldingAmount } from "./TokenHoldingAmount.sol";
+import { Btcs } from "./brc20/Btcs.sol";
+import { Cats } from "./brc20/Cats.sol";
```
Comment (BillyWooo) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol]: only these two? I see there are quite some others, like long, mmss, ordi, rats, sats?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol [hunk 2]
```diff
@@ -21,62 +21,86 @@ pragma solidity ^0.8.8;
...
+		string memory tokenBscAddress = "";
+		string memory tokenEthereumAddress = "";
+
+		if (Utils.isStringsEqual(decodedParams, "Btcs")) {
+			tokenName = Btcs.getTokenName();
+			ranges = Btcs.getTokenRanges();
+		} else if (Utils.isStringsEqual(decodedParams, "Cats")) {
+			tokenName = Cats.getTokenName();
+			ranges = Cats.getTokenRanges();
+		} else {
```
Comment (BillyWooo) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol]: same problem as above. Just forget?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol [hunk 3]
```diff
@@ -17,34 +17,40 @@
...
 pragma solidity ^0.8.8;
+
 import "../libraries/Identities.sol";
 import "../libraries/Utils.sol";
 import { TokenHoldingAmount } from "./TokenHoldingAmount.sol";
 import { NoderealClient } from "./NoderealClient.sol";
-abstract contract ERC20 is TokenHoldingAmount {
-	mapping(uint32 => string) internal networkTokenAddresses;
+import { Ada } from "./erc20/Ada.sol";
+import { Amp } from "./erc20/Amp.sol";
```
Comment (BillyWooo) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol]: similar problem as above. 
Do I misunderstand anything?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol [hunk 4]
```diff
@@ -96,4 +102,34 @@ abstract contract ERC20 is TokenHoldingAmount {
...
+			tokenName = Ada.getTokenName();
+			ranges = Ada.getTokenRanges();
+			tokenBscAddress = Ada.getTokenBscAddress();
+			tokenEthereumAddress = Ada.getTokenEthereumAddress();
+		} else if (Utils.isStringsEqual(decodedParams, "Amp")) {
+			tokenName = Amp.getTokenName();
+			ranges = Amp.getTokenRanges();
+			tokenBscAddress = Amp.getTokenBscAddress();
+			tokenEthereumAddress = Amp.getTokenEthereumAddress();
+		} else {
```
Comment (BillyWooo) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol]: same here?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol (hunk 3 above)
Comment (0xverin) [reply tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol]: I added all and made some changes to the positions.

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/brc20/BRC20TokenLibrary.sol [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import { Btcs } from "./Btcs.sol";
+import { Cats } from "./Cats.sol";
+import { Long } from "./Long.sol";
+import { Mmss } from "./Mmss.sol";
+import { Ordi } from "./Ordi.sol";
+import { Rats } from "./Rats.sol";
+import { Sats } from "./Sats.sol";
+
+library BRC0TokenLibrary {
+	function getBtcsInfo()
```
Comment (wli-pro) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/brc20/BRC20TokenLibrary.sol]: Would mapping {string => uint256[] ranges} work, i.e. we will  have a mapping (somewhat hard coded, but simplier) like:
{
"sats": Sats.getTokenRanges(),
"rats": Rats.getTokenRanges(),
}

then we can get rid of the hard coded getXyzInfo, instead use the mapping dynamicly map tokenName to get the token's range?

**Comment by wli-pro** [comment]
@0xverin Ok I might have potentially found one more gap: the generated VC is lack of the $network info, which is for what networks the VC calculated on, see example https://github.com/litentry/litentry-parachain/pull/2673

**Comment by 0xverin** [comment]
> @0xverin Ok I might have potentially found one more gap: the generated VC is lack of the $network info, which is for what networks the VC calculated on, see example #2673

I see in #2785  it has already changed. If fix are needed, shall we address them in the new PR?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/brc20/BRC20TokenLibrary.sol (hunk 5 above)
Comment (0xverin) [reply tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/brc20/BRC20TokenLibrary.sol]: like this?:
```contract ERC20Mapping is ERC20{
    constructor(){
        // ada
        tokenNames['ada'] = Ada.getTokenName();
        tokenRanges['ada'] = Ada.getTokenRanges();
        tokenBscAddress['ada'] = Ada.getTokenBscAddress();
        tokenEthereumAddress['ada'] = Ada.getTokenEthereumAddress();
        // amp
        tokenNames['amp'] = Amp.getTokenName();
	tokenNames['amp'] = Amp.getTokenRanges();
	tokenNames['amp'] = Amp.getTokenBscAddress();
	tokenNames['amp'] = Amp.getTokenEthereumAddress();
}

Comment (0xverin) [reply tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/brc20/BRC20TokenLibrary.sol]: Did some changes.

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/libraries/Constants.sol [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.8;
+
+library Constants {
+	uint256 constant decimals_factor = 1000;
```
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/libraries/Constants.sol]: This is only related to token holdings so please move this to `token_holding_amount` directory.

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol [hunk 7]
```diff
@@ -22,33 +22,65 @@ import "https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.0/contr
...
+			string memory,
+			bool
+		)
+	{
+		string
+			memory description = "The amount of a particular token you are holding";
+		string memory assertion_type = "Token Holding Amount";
+		schema_url = "https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-0.json";
+
+		string memory decodedParams = abi.decode(params, (string));
```
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol]: is this token? Could you rename variable ?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol [hunk 8]
```diff
@@ -22,33 +22,65 @@ import "https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.0/contr
...
+
+		if (
+			keccak256(abi.encodePacked(tokenNames[decodedParams])) ==
+			keccak256(abi.encodePacked(""))
+		) {
+			revert("Token not supported or not found");
+		}
+
+		tokenAddresses[Web3Networks.Bsc] = tokenBscAddress[decodedParams];
```
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol]: Can this be done in `BRC20.sol` ?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol [hunk 9]
```diff
@@ -22,33 +22,65 @@ import "https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.0/contr
...
-            memory description = "The amount of a particular token you are holding";
-        string memory assertion_type = "Token Holding Amount";
-        schema_url = "https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-0.json";
-
-		uint256 balance = queryTotalBalance(identities, secrets);
-
-		(uint256 index, uint256 min, int256 max) = calculateRange(balance);
-
-		string[] memory assertions = assembleAssertions(min, max);
+	mapping(uint32 => string) tokenAddresses;
```
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol]: Can we have these mappings only for `ERC20` and `BRC20`?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol [hunk 10]
```diff
@@ -21,62 +21,59 @@ pragma solidity ^0.8.8;
 import "../libraries/Identities.sol";
 import "../libraries/Http.sol";
 import "../libraries/Utils.sol";
-import {TokenHoldingAmount} from "./TokenHoldingAmount.sol";
+import { TokenHoldingAmount } from "./TokenHoldingAmount.sol";
+contract BRC20 is TokenHoldingAmount {
```
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol]: Can this be abstract ? 

Is this code meant to be uploaded or `BRC20Mapping.sol` ?

**Review by kziemianek — CHANGES_REQUESTED** [review]
I though we will have one instance for all token holdings, after reading this PR it turns out that's not the case. It's ok if this is the design you and @higherordertech prefer.

Despite the comments, byte code in repo should be updated so we can see tests pass with bytecode produced from new solidity code.
https://github.com/litentry/litentry-parachain/pull/2811/files#diff-afbca31fd80a6d91ed61556c33e9b8c6e61f6df4b46fa9a91d75b21afb69ac67

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol (hunk 8 above)
Comment (0xverin) [reply tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/TokenHoldingAmount.sol]: Done.

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol (hunk 10 above)
Comment (0xverin) [reply tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/BRC20.sol]: fixed.

**Comment by 0xverin** [comment]
> I though we will have one instance for all token holdings, after reading this PR it turns out that's not the case. It's ok if this is the design you and @higherordertech prefer.
> 
> Despite the comments, byte code in repo should be updated so we can see tests pass with bytecode produced from new solidity code. https://github.com/litentry/litentry-parachain/pull/2811/files#diff-afbca31fd80a6d91ed61556c33e9b8c6e61f6df4b46fa9a91d75b21afb69ac67

I have not merged brc20 and erc20 temporarily, because:

1. It involves the refactoring of querybalance, and I still don't know the structure of btc solana for now. I want to wait until they are all completed before merging all.

2. Even if we can merge them all, I am worried that it will still affect the use of contracts, as there have been compilation warnings currently:

```
Warning: Contract code size is 38662 bytes and exceeds 24576 bytes (a limit introduced in Spurious Dragon). This contract may not be deployable on mainnet. Consider enabling the optimizer (with a low "runs" value!), turning off revert strings, or using libraries.
  --> contracts/token_holding_amount/erc20/ERC20Mapping.sol:63:1:
  ```

Wdyt? How about we wait for other tokens to be added before merging them together?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/libraries/Constants.sol
Comment (Kailai-Wang) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/libraries/Constants.sol:1]: duplicate of `tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/Constants.sol`?

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol [hunk 11]
```diff
@@ -17,47 +17,59 @@
...
+		] = "https://eth-mainnet.nodereal.io/v1/";
 		// Add more networks as needed
-		
-		// 	below url is used for test against mock server
+		// below url is used for test against mock server
 		// "http://localhost:19530/nodereal_jsonrpc/v1/",
+
+		// Initialize network token addresses using Ada library
+
+
```
Comment (Kailai-Wang) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/ERC20.sol]: Empty comment, or placeholders?

**Review by Kailai-Wang — COMMENTED** [review]
Questions:
1. imagine the client adds a new token, does it mean that the `ERC20Mapping.sol` needs to be extended and reployed?
2. as more and more tokens are added, this all-in-one contract will become bigger and bigger. Would it be actually lighter to deploy a separate `tokenX.sol` which inherits the ERC20 contract (so similar to what people do on ethereum).

I'm just tthinking it can be challenging to mantain the all-in-one contract

**Comment by wli-pro** [comment]
> Questions:
> 
> 1. imagine the client adds a new token, does it mean that the `ERC20Mapping.sol` needs to be extended and reployed?
> 2. as more and more tokens are added, this all-in-one contract will become bigger and bigger. Would it be actually lighter to deploy a separate `tokenX.sol` which inherits the ERC20 contract (so similar to what people do on ethereum).
> 
> I'm just tthinking it can be challenging to mantain the all-in-one contract

What's the plan in my head is:
1 for initial tokens we migrated we prefer one contract for all, simply because the easier deployment for them.
2 none of our solidity VC is in production usage yet, so this one contract for all way is easier if we find contract defects and we will need fix which need redeploy all of relevant tokens. If the common code is stable, proven defects free, then one contract per token would probably be easier. Before that as it still evolves one contract for all token could be easier to maintain.
3 for new token to be added in future, it's not enforced either way, if VC maintainer feels it's easier for them to add in this one contract or if they feel they want to modify this all token contract to one token contract (that should be easy to do too, or we can provide a placeholder version for that) they can do either of it.

**Comment by 0xverin** [comment]
Updates:

1. Merged ERC20 and BRC20, we only need to deploy the TokenMapping contract

2. Updated dynamic mod test cases, also updated the bytecode of repository

cc @litentry/parachain 
Test evidence:
```
./bin/litentry-cli trusted -d link-identity did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e did:litentry:evm:0x7001f26ebc3e70609736a82ac3f2c8208ffa75ca bsc
 ```
 ```
  ./bin/litentry-cli trusted -d request-vc did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e -a "dynamic 0000000000000000000000000000000000000001 00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000003626e620000000000000000000000000000000000000000000000000000000000"
  ```
  
  ```
  {
    "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/suites/ed25519-2020/v1"
    ],
    "id": "0xa6e460efcbb9b4e130f6e761030cfbc20671ca36168b2b79cf53eaed8526d434",
    "type": [
        "VerifiableCredential"
    ],
    "credentialSubject": {
        "id": "did:litentry:substrate:0xe659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e",
        "description": "The amount of a particular token you are holding",
        "type": "Token Holding Amount",
        "assertionText": "Dynamic(0x0000000000000000000000000000000000000001, BoundedVec([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 98, 110, 98, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 1024))",
        "assertions": [
            {
                "and": [
                    {
                        "src": "$token",
                        "op": "==",
                        "dst": "bnb"
                    },
                    {
                        "src": "$holding_amount",
                        "op": ">=",
                        "dst": "800"
                    },
                    {
                        "src": "$holding_amount",
                        "op": "<",
                        "dst": "1200"
                    }
                ]
            }
        ],
        "values": [
            true
        ],
        "endpoint": "http://localhost:9933/"
    },
    "issuer": {
        "id": "did:litentry:substrate:0x7ccfdef3e7cd7bc5bac3b5845095e13c98f7eee5fd481884c7398d877d792b15",
        "name": "Litentry TEE Worker",
        "mrenclave": "2GTxBHhQQyGwYXHRkRAkpchZAmvEJ1JGUw5Xr15Wc11F",
        "runtimeVersion": {
            "parachain": 9181,
            "sidechain": 107
        }
    },
    "issuanceDate": "2024-06-21T05:42:33.614549656+00:00",
    "parachainBlockNumber": 344,
    "sidechainBlockNumber": 681,
    "proof": {
        "created": "2024-06-21T05:42:33.615823708+00:00",
        "type": "Ed25519Signature2020",
        "proofPurpose": "assertionMethod",
        "proofValue": "e9eea72402a3de32a356f4984339b2b5653fa499936a5cffd1f093cf9b8fda5f1fd64fb9a21ffb084a546f377628da65d60f256824acbc980c104b296ade0e0c",
        "verificationMethod": "0x7ccfdef3e7cd7bc5bac3b5845095e13c98f7eee5fd481884c7398d877d792b15"
    },
    "credentialSchema": {
        "id": "https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-0.json",
        "type": "JsonSchemaValidator2018"
    }
}
```

image

**Review by BillyWooo — APPROVED** [review]
LGTM
Let's roll it out at least for dev env and test it.

**Review by wli-pro — APPROVED** [review]
Approved this on condition that there is no issue to deploy it to parachain (accept that it might have issues in remix)

**Review by kziemianek — APPROVED** [review]

**Comment by wli-pro** [comment]
This PR doesn't make rust interface change but only adjusted / refactored the solidity VC implementation, hence it's not a breaking change to consumer application, hence remove breaking change tag unless any other objection raised.

**Pull request merged**